cwl/src/main/scala/cwl/CwlCodecs.scala

```diff
@@ -22,5 +31,44 @@ object CwlCodecs {
   implicit val cltD = implicitly[Decoder[CommandLineTool]]
   implicit val etD = implicitly[Decoder[ExpressionTool]]
 
-  def decodeCwl(in: String): Checked[CwlFile] = decodeAccumulating[CwlFile](in).leftMap(_.map(_.getMessage).map(s"error parsing: $in" + _)).toEither
+  def errorToStrings : NonEmptyList[Error] => NonEmptyList[String] = _.map(_.show)
+
+  def decodeWithErrorStrings[A](in: String)(implicit d: Decoder[A]): ErrorOr[A] = decodeAccumulating[A](in).leftMap(errorToStrings)
+
+  /**
+    * Attempt to parse as an array of CWL or a single one.
+    *
+    * If it fails, attempt to parse each CWL individually and return a superset of those failures.
+    */
+  def decodeCwl(in: String): Checked[CwlFile] =
+    decodeWithErrorStrings[CwlFile](in).leftMap(decodePieces(in)).toEither
+
+  /**
+    * Drop down to a lower level and try to parse each CWL individually, returning a superset of those failures.
+    */
+  def decodePieces(in: String)(original: NonEmptyList[String]):NonEmptyList[String]  =
+    (for {
+      raw <- parse(in).leftMap(_.show).leftMap(NonEmptyList.one)
+      _ <-
+        if (raw.isArray)
+          raw.asArray.toVector.flatten.traverse(decodeCwl).toEither
+        else
+          decodeCwl(raw).toEither
+    } yield  ()).fold(
+      identity,
+      _ => original //this should never happen as we're gathering up the errors
+    )
+
+  private def decodeCwl(json: Json): ErrorOr[Unit] = {
+
+    val rawJson = io.circe.Printer.noSpaces.pretty(json)
+
+    // we know we are dealing with an individual CWL, so we can take some liberties like assuming "class" is defined.
+    ((json \\ "class").head.asString match {
+      case Some("Workflow") => decodeWithErrorStrings[Workflow](rawJson)
+      case Some("CommandLineTool") => decodeWithErrorStrings[CommandLineTool](rawJson)
+      case Some("ExpressionTool") => decodeWithErrorStrings[ExpressionTool](rawJson)
+      case _ => "Class field was declared incorrectly!".invalidNel
```

Comment:
The message could be more helpful, possibly including the actual unacceptable value.